### PR TITLE
amd gpu ep for multi-stream work

### PR DESCRIPTION
**Pull request by jslhcl**
### Description
amd gpu ep for multi-stream work

### Motivation and Context
amd gpu ep for multi-stream work

### CODE REVIEW

File: onnxruntime/core/providers/migraphx/gpu_data_transfer.cc [hunk 1]
```diff
@@ -6,56 +6,48 @@
...
+//    // copying between cpu memory
+//    memcpy(dst_data, src_data, bytes);
+//  }
+//
+//  return Status::OK();
+//}
+
+common::Status GPUDataTransfer::CopyTensorAsync(const Tensor& /*src*/, Tensor& /*dst*/, Stream* /*stream*/) const {
+  // TODO:
   return Status::OK();
```
Comment (souptc) [inline onnxruntime/core/providers/migraphx/gpu_data_transfer.cc:51]: ? it is not implemented? how could it work?

File: onnxruntime/core/providers/migraphx/migraphx_execution_provider.h [hunk 2]
```diff
@@ -54,9 +54,10 @@ class MIGraphXExecutionProvider : public IExecutionProvider {
 
   void RegisterAllocator(AllocatorManager& allocator_manager) override;
 
-  void* GetComputeStream() const override { return static_cast<void*>(stream_); }
+  void* GetComputeStream() const { return static_cast<void*>(stream_); }  // TODO: this function should be deleted
```
Comment (souptc) [inline onnxruntime/core/providers/migraphx/migraphx_execution_provider.h:57]: why you still keep it? this API should be removed.

File: onnxruntime/core/providers/rocm/math/gemm.cc [hunk 3]
```diff
@@ -106,7 +106,7 @@ Status Gemm<T>::ComputeInternal(OpKernelContext* ctx) const {
           N, M, 1,
           /*alpha*/ &one,
           b_data, N,
-          GetConstOnes<HipT>(M), 1,
+          GetConstOnes<HipT>(M, GetRocmStreamFromContext(ctx)), 1,
```
Comment (souptc) [inline onnxruntime/core/providers/rocm/math/gemm.cc:109]: i thought we have the API Stream(ctx)?

File: onnxruntime/core/providers/rocm/rocm_execution_provider.cc [hunk 4]
```diff
@@ -2406,4 +2408,27 @@ void ROCMExecutionProvider::RegisterAllocator(AllocatorManager& allocator_manage
   }
 }
 
+void ROCMExecutionProvider::RegisterStreamHandlers(IStreamCommandHandleRegistry& stream_handle_registry) const {
+  // This allocator must be the same to the allocator
+  // used in AllocateBufferOnCPUPinned.
+  auto allocator = GetAllocator(DEFAULT_CPU_ALLOCATOR_DEVICE_ID, OrtMemTypeCPU);
+  if (use_ep_level_unified_stream_)
```
Comment (souptc) [inline onnxruntime/core/providers/rocm/rocm_execution_provider.cc:2415]: the if-else looks the same?

File: onnxruntime/core/providers/rocm/rocm_execution_provider.h [hunk 5]
```diff
@@ -17,6 +17,8 @@
 
 namespace onnxruntime {
 
+void RunOnUnload(std::function<void()> function);
```
Comment (souptc) [inline onnxruntime/core/providers/rocm/rocm_execution_provider.h:20]: why you add this? is it hipfied?

File: onnxruntime/core/providers/rocm/rocm_kernel.h [hunk 6]
```diff
@@ -58,14 +60,23 @@ class RocmKernel : public OpKernel {
     return provider_->AllocateBufferOnCPUPinned<T>(count_or_bytes);
   }
 
-  inline void AddDeferredReleaseCPUPtr(void* p) const {
+  inline void AddDeferredReleaseCPUPtr(void* p, onnxruntime::Stream* /*TODO*/= nullptr) const {
```
Comment (souptc) [inline onnxruntime/core/providers/rocm/rocm_kernel.h:63]: why you add this? have you checked does this work with the old "deferred release CPU ptr" approach?

File: onnxruntime/core/providers/rocm/rocm_kernel.h [hunk 7]
```diff
@@ -105,6 +116,11 @@ class RocmKernel : public OpKernel {
       return Status::OK();
     }
 
+    Status CopyToGpu(onnxruntime::Stream* /*stream*/) {
+      // TODO:
```
Comment (souptc) [inline onnxruntime/core/providers/rocm/rocm_kernel.h:120]: again, why add this?

File: onnxruntime/core/providers/shared_library/provider_interfaces.h [hunk 8]
```diff
@@ -171,7 +171,7 @@ struct ProviderHost {
 #ifdef USE_ROCM
   virtual std::unique_ptr<IAllocator> CreateROCMAllocator(int16_t device_id, const char* name) = 0;
   virtual std::unique_ptr<IAllocator> CreateROCMPinnedAllocator(int16_t device_id, const char* name) = 0;
-  virtual std::unique_ptr<IDataTransfer> CreateGPUDataTransfer(void* stream) = 0;
```
Comment (souptc) [inline onnxruntime/core/providers/shared_library/provider_interfaces.h:174]: why change the shared provider api?

File: onnxruntime/core/providers/migraphx/migraphx_execution_provider.h (hunk 2 above)
Comment (TedThemistokleous) [reply onnxruntime/core/providers/migraphx/migraphx_execution_provider.h:57]: No it shouldn't MIGraphX requires this to operate.

**Pull request closed without merging**